rfcs/0008-custom-event-metadata-from-annotations/README.md

```diff
@@ -0,0 +1,219 @@
+Flux objects rather than introducing a new API.
+
+In the specific case of the Alert API this field was introduced because the Alert API
+is obviously a special one in the context of events and alerting. In particular, the
+Alert objects do not generate events themselves, but rather serve as an aggregation
+configuration for matching and propagating events from other Flux objects.
+
+#### Alternative 2
+
+Instead of introducing a new API Group, i.e. `event.toolkit.fluxcd.io`, we could use the API
+Group `notification.toolkit.fluxcd.io` for the same purpose. This alternative is not great
+because it emphasizes an exclusive relationship with notification-controller, which is not the
+case. The events here are also Kubernetes Events, and an API Group that is more general and
+closer to Kubernetes Events is more appropriate.
+
+## Design Details
+
+All the Flux controllers use our implementation of the `EventRecorder` interface from the Go
+package `k8s.io/client-go/tools/record`: [`(*github.com/fluxcd/pkg/runtime/event.Recorder).AnnotatedEventf()`](https://github.com/fluxcd/pkg/blob/6f2619522699f1a78e8c7b41583ad9f7b7c9544e/runtime/events/recorder.go#L119).
+This implementation sends the events to notification-controller and also calls the same method
+from an injected `EventRecorder`, which sends the events to Kubernetes. To support the use cases
+discussed here we would modify this implementation to look for annotations prefixed with
+`event.toolkit.fluxcd.io/` in the Flux objects and send them alongside the other metadata of
+the event. Here we are talking specifically about the object annotations retrieved from the
+Flux object itself, i.e. the first argument of the `AnnotatedEventf()` method: `object runtime.Object`.
+This implementation would not change the interface `EventRecorder` used by the controllers,
+so all we need to do is bump the Go package `github.com/fluxcd/pkg/runtime` across all controllers.
+
+On the notification-controller side we would start accepting metadata keys starting with this
+prefix and remove it before sending the metadata key-value pair to the notification providers.
+This is an important aspect of the implementation because notification-controller only
+accepts metadata keys that are prefixed with the Group of the respective API the involved
+Flux object belongs to, so we need to add an exception for the new prefix.
+
+After this change, there would be four sources of metadata being sent by notification-controller
+to the providers. They are listed below with the proposed order of precedence, from lowest to highest:
+
+1. User-defined metadata on objects (set with the `event.toolkit.fluxcd.io/` prefix in the event payload)
+2. User-defined metadata on the alert (set with `.spec.eventMetadata`)
+3. User-defined summary on the alert (set with `.spec.summary`)
+4. Controller metadata (set with the `<object group>.toolkit.fluxcd.io/` prefix in the event payload)
```

Comment:
I think this should be the other way around, user-defined metadata on objects override the "global" ones set in the Alert.

Reply:
fixed according to our new agreement on the other thread, thanks